File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBStValid=true;
    void dfs(TreeNode *root, TreeNode *lower, TreeNode *higher){
        if(root==NULL) return ;
        dfs(root->left,lower,root);
        if((lower &&  root->val<=lower->val) || (higher && root->val>=higher->val)){
            isBStValid=false;;
        }
        dfs(root->right,root,higher);
        
    }
    bool isValidBST(TreeNode* root) {
        if(root==NULL) return true;
        dfs(root,NULL,NULL);
        return isBStValid;
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValid=true;
    public boolean dfs(TreeNode root, TreeNode lower, TreeNode upper){
        if(root==null) return true;
        if(lower!=null){
            if(root.val<=lower.val) return false;
        }
        if(upper!=null){
            if(root.val>=upper.val) return false;
        }
        return dfs(root.left,lower,root) && dfs(root.right,root,upper);
    }
    public boolean isValidBST(TreeNode root) {
        if(root==null) return true;
        return dfs(root,null,null);
        
    }
}